src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs

```diff
@@ -35,20 +36,59 @@ public override string Run(TaskContext context, string input)
 
         public async override Task<string> RunAsync(TaskContext context, string rawInput)
         {
-            HttpRequestMessage requestMessage = await this.ReconstructHttpRequestMessage(rawInput);
-            HttpResponseMessage response = await this.httpClient.SendAsync(requestMessage);
-            DurableHttpResponse durableHttpResponse = await DurableHttpResponse.CreateDurableHttpResponseWithHttpResponseMessage(response);
+            DurableHttpRequest durableHttpRequest = ReconstructDurableHttpRequest(rawInput);
+            HttpRequestMessage requestMessage = await this.ReconstructHttpRequestMessage(durableHttpRequest);
 
-            return JsonConvert.SerializeObject(durableHttpResponse);
+            CancellationTokenSource cts = new CancellationTokenSource();
+            if (durableHttpRequest.Timeout != null)
+            {
+                try
+                {
+                    TimeSpan timeout = durableHttpRequest.TimeoutExpirationDateTime - DateTime.UtcNow;
+                    this.httpClient.Timeout = timeout;
+                    cts.CancelAfter(timeout);
+                }
+                catch (Exception)
+                {
+                    TimeSpan timeout = durableHttpRequest.Timeout.Value;
+                    this.httpClient.Timeout = timeout;
+                    cts.CancelAfter(timeout);
+                }
```

Comment:
Any suggestions on what to do if `DateTime.UtcNow` is greater than `durableHttpRequest.TimeoutExpirationDateTime`? Right now, it catches the exception that says that the `timeout` TimeSpan is invalid and resets the `timeout` value. Should it just throw a TimeoutException?

Reply:
So I can see an argument for one of two approaches:

1. When the customer sets the `timeout` property, it is absolute. If it expires before we hit the activity, don't even bother with the HTTP request. The good part of this approach is that it is straightforward to explain how it works. There are a couple of downsides. The big one is that the request could never even fire successfully due to internal Durable Task latency (such as a queue backlog). The other downside I see is that it doesn't give the customer much more usefulness other than just doing a Task.WhenAny() with a Timer task.

2. We only check for the timeout before we "schedule" the HTTP call. The downside is it is unclear what to set as the request cancellation timeout token to if the timeout has expired when we enter the HttpActivityShim.

Reply:
@cgillum, what are your thoughts here?

Reply:
I'm not actually sure I understand the difference between those two options. However, it looks like we need to make another design decision about how the timeout should work. It seems there are two options, and maybe you were already alluding to this and I misunderstood.

1. Timeout is "absolute" in the sense that it represents a point in time. In this model, it would also include time spent in the work-item queue.
2. Timeout is relative in the sense that it does not have a specific expiration, but rather dictates how much time the HTTP client should wait for a response.

My gut tells me that (2) will be a more common scenario. In that case, we don't need to calculate any expiration time at all. We just pass the timeout value all the way to this point and use it in the call to `cts.CancelAfter()`. I worry that with (1) customers will have unpredictable and/or unexpected runtime behavior.

Thoughts?

Reply:
I think it will be easier to document and maintain the second option (the timeout specifies the time the HTTP client should wait for a response). I agree that option 1 might be confusing for customers.